core/src/kvs/scanner.rs

```diff
@@ -0,0 +1,108 @@
+use super::tx::Transaction;
+use super::Key;
+use super::Val;
+use crate::err::Error;
+use futures::stream::Stream;
+use futures::Future;
+use futures::FutureExt;
+use std::collections::VecDeque;
+use std::ops::Range;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+type Output = Result<Vec<(Key, Val)>, Error>;
+
+pub(super) struct Scanner<'a> {
+	/// The store which started this range scan
+	store: &'a Transaction,
+	/// The number of keys to fetch at once
+	batch: u32,
+	// The key range for this range scan
+	range: Range<Key>,
+	// The results from the last range scan
+	results: VecDeque<(Key, Val)>,
+	/// The currently running future to be polled
+	future: Option<Pin<Box<dyn Future<Output = Output> + 'a>>>,
+	/// Whether this stream should try to fetch more
+	exhausted: bool,
+}
+
+impl<'a> Scanner<'a> {
+	pub fn new(store: &'a Transaction, batch: u32, range: Range<Key>) -> Self {
+		Scanner {
+			store,
+			batch,
+			range,
+			future: None,
+			results: VecDeque::new(),
+			exhausted: false,
+		}
+	}
+}
+
+impl<'a> Stream for Scanner<'a> {
+	type Item = Result<(Key, Val), Error>;
+	fn poll_next(
+		mut self: Pin<&mut Self>,
+		cx: &mut Context<'_>,
+	) -> Poll<Option<Result<(Key, Val), Error>>> {
+		// If we have results, return the first one
+		if let Some(v) = self.results.pop_front() {
+			return Poll::Ready(Some(Ok(v)));
+		}
+		// If we won't fetch more results then exit
+		if self.exhausted {
+			return Poll::Ready(None);
+		}
+		// Set the max number of results to fetch
+		let num = std::cmp::min(1000, self.batch);
+		// Clone the range to use when scanning
+		let range = self.range.clone();
```

Comment:
I think we should move these two into the `if` statement below as we only need them for constructing the future.

Reply:
Good spot @rushmorem 👍 ! Fixed in [581751e](https://github.com/surrealdb/surrealdb/pull/4257/commits/581751e5ff585a9866796da72a20bf6adc465eac).